### PR TITLE
background tasks in Python using task queues

### PR DESCRIPTION
**Pull request by divyanshutomar**
Article on running background tasks in Python using task queues

### CODE REVIEW

File: posts/running-background-tasks-python.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
+---
+title: Running background tasks in Python using task queues
+author: Divyanshu Tomar
+---
+# Running background tasks in Python using task queues
+
+We often come across problems in our applications where a compute-intensive time-taking task needs to be performed on the server in response to some user activity or request. For a server-side application exposing a REST API, handling this problem is different from common CRUD endpoints where request-response lifecycle is usually short. In this case, the response for such a request may not be available, as it may be not viable to proceed with its execution immediately on the same process.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: Rather than starting with 'We often come across problems...", it would be more powerful to give an example where this occurs.

Comment (sidhanthp) [reply posts/running-background-tasks-python.md]: Additionally, I think this example is a little convoluted for the 1st paragraph. 

For a server-side application exposing a REST API, handling this problem is different from common CRUD endpoints where request-response lifecycle is usually short. In this case, the response for such a request may not be available, as it may be not viable to proceed with its execution immediately on the same process.
-->
When dealing with server-side applications, the response from a REST API may not be available immediately. It would be more efficient to run background tasks than waste the idle CPU cycles.

File: posts/running-background-tasks-python.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+The execution of such tasks or jobs can be performed in the background by some another process spawned for the sole purpose. These processes are usually called _workers_. They run concurrently with the main process (web server in our case) handling client requests. To list of all the tasks which need to be executed, a job/task queue is maintained to store tasks along with its metadata created by incoming requests on the web server. The worker process then executes these tasks chronologically. This modular approach makes it easier for the web server to accommodate execution of such long-running tasks as it will not get blocked itself in doing so. This also means that the web server can respond to forthcoming client requests.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I assume you meant to write "by another process".

Comment (sidhanthp) [reply posts/running-background-tasks-python.md]: "To list all the tasks"

File: posts/running-background-tasks-python.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+![architectire of web server and queue](./images/running-background-tasks-python/small-archi.png)
+
+Task queues are quite popular among microservices architecture. They enable each microservice to perform its definite task really well and take care of the complexities of inter-microservice communication.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I'm not sure what you mean when you say 'definite task'. If I were to guess, you're referring to the idea that each microservice should perform a small task.

Additionally, if you make a claim that a task queue takes 'care of the complexities of inter-microservice communication', I would explain it a little further (unless I'm missing something).

File: posts/running-background-tasks-python.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+```
+git clone https://github.com/divyanshutomar/flask-task-queue.git
+```
+
+To check if everything is working fine, start Docker daemon and run `docker-compose up --build` to start the application. This launches two containers, our web server application and Redis server interlinked with each other using Docker networking. The `--build` argument makes sure the image is built using the latest code every time while running the containers. Visit `localhost:5000` in a browser to check if web service is working fine. This is the default URL that Gunicorn runs our flask application on.
+
+![health check](./images/running-background-tasks-python/health-check.png)
+
+
+### Getting to Know Starter Application
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I would use `this` to make the tool names stand out, rather than *this*

File: posts/running-background-tasks-python.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+![services cloud](./images/running-background-tasks-python/lib-cloud.png)
+
+*Pipenv* is used for managing the python dependencies and environment. We use *Docker* for containerising and running our application in an isolated environment. Dockerfile lets us specify how our application should be run. Docker compose lets us orchestrate multiple containers (flask web server and Redis server in our case) to run together and be able to communicate with each other.
+
+### Writing the parser
+
+Let's start with writing a simple parser that accepts a Goodreads book page URL. We will be using requests python library for making an HTTP request to get HTML content of the page. BeautifulSoup is a python library that lets us search, manipulate and create structured markup languages such as HTML, XML, etc. It will create a searchable tree from the fetched page's HTML. This will allow us to retrieve key information like the book title, author, rating, and description.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I would leave out "It will create a searchable tree from the fetched page's HTML."

File: posts/running-background-tasks-python.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+### [Redis Queue](http://python-rq.org/)
+
+Redis Queue (RQ) is a python library which implements a simple job queue using Redis. It's a lightweight alternative that derives the good parts from existing queueing frameworks. A single worker process can be pointed to many queues to execute the tasks. We will be using this in the following application for its simplicity.
+
+
+There are many other queueing frameworks or services available. Refer [queues.io](http://queues.io/) to discover more.
+
+## Real World Application
+
+We will be writing a flask-based web application which uses Redis Queue (RQ) python library for executing asynchronous tasks. The web server exposes an endpoint that accepts Goodreads book URLs. A function will parse this URL for meta information of the book like title, author, rating, etc. The function will execute asynchronously by a separate worker process. It requires a Redis server as a message broker for performing this operation. Let's get into the code and learn how we can use Redis queue in our web applications.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I would include a brief explanation of why this is a good case for using RQ, since the time the process would be waiting for a response from the server is significantly more than the time spent processing.

File: posts/running-background-tasks-python.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+```
+
+We can now write a function that calls the above parsing function and persists the value to Redis so that it can be retrieved later. This function along with its arguments will be pushed to queue so that the worker process can execute it.
+
+```python
+# server.py
+
+generate_redis_key_for_book = lambda bookURL: 'GOODREADS_BOOKS_INFO:' + bookURL
+
+def parse_and_persist_book_info(bookUrl):
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I wouldn't say I'm left with a strong understanding of Redis and what it can do. If this is the bulk of your code surrounding Redis, I would go more in-depth and explain what you're doing, why you're doing it, and what's happening internally. 

It feels like I'm just copying & pasting code w/o understanding it. Not a bad idea when we're working on the web scraping, but if we're trying to teach a person how to Redis Queue's, we should explain what's going on when we're writing the code.

File: posts/running-background-tasks-python.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+# Endpoint that accepts an array of Goodreads URLs for meta information parsing
+@app.route('/parseGoodreadsLinks', methods=["POST"])
+def parse_goodreads_urls():
+  bodyJSON = request.get_json()
+  if (isinstance(bodyJSON,list) and len(bodyJSON)):
+    bookLinksArray = [x for x in list(set(bodyJSON)) if x.startswith('https://www.goodreads.com/book/show/')]
+    if (len(bookLinksArray)):
+      for bookUrl in bookLinksArray:
+        bookInfoParserQueue.enqueue_call(func=parse_and_persist_book_info,args=(bookUrl,),job_id=bookUrl)
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: Explain this further.

File: posts/running-background-tasks-python.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+```python
+# server.py
+
+# Endpoint for retrieving book info from Redis
+@app.route('/getBookInfo', methods=["GET"])
+def get_goodreads_book_info():
+  bookURL = request.args.get('url', None)
+  if (bookURL and bookURL.startswith('https://www.goodreads.com/book/show/')):
+    redisKey = generate_redis_key_for_book(bookURL)
+    cachedValue = redisClient.get(redisKey)
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: Explain what's going on here further.

File: posts/running-background-tasks-python.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    if cachedValue:
+      return jsonify(pickle.loads(cachedValue))
+    return "No meta info found for this book."
+  return "'url' query parameter is required. It must be a valid goodreads book URL.",400
+```
+
+### Inspecting task queue
+
+The creators of Redis queue (RQ) library have developed another library for checking the state of Redis queue. It is called *rq-dashboard* and it can be integrated with our flask web application. It exposes a flask blueprint for integrating with an existing flask project. It's a browser-based application which shows queues status,  workers listening on those queues and jobs queued along with their meta information. Also, it provides triggers for flushing the queue and re-queuing failed jobs.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I think you can make this a little more concise and give a brief description of the benefits.

File: posts/running-background-tasks-python.md [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+![rq dashboard](./images/running-background-tasks-python/rq-dash.png)
+
+### Testing the application
+
+Now, we are all set to begin testing our application with some Goodreads URLs. Let's start by making a POST request to `/parseGoodReads` endpoint. Make sure to provide a valid list of URLs in an array as the request body.
+
+![post request](./images/running-background-tasks-python/post-req.png)
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: What are you using to make the post request?

File: posts/running-background-tasks-python.md (hunk 10 above)
Comment (sidhanthp) [reply posts/running-background-tasks-python.md]: It looks like you've written about it a little below, I think it might be better to move that up.

File: posts/running-background-tasks-python.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+![get book request](./images/running-background-tasks-python/get-book-info.png)
+
+You can view the full code [here](https://github.com/divyanshutomar/flask-task-queue/tree/completed).
+
+## Conclusion and Takeaways
+
+The above application demonstrates how queuing frameworks like Redis Queue can be leveraged for solving problems that need more time and computing resources. In our application, if the parsing task is executed on the same process where client requests are being processed and served, it can easily become a performance bottleneck with high traffic. The above approach not only helps in avoiding that but also bring modularity to the table. Following are some of the key takeaways you can follow to tackle similar problems:
+* Queueing frameworks allows more granular control over scaling of different processes. More worker processes can be spawned if there is an accumulation of a large number of tasks in the queue.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: "frameworks allow more"

File: posts/running-background-tasks-python.md [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+* Multiple queues can be used for handling different type of tasks.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: Did you talk about this earlier in the post?

File: posts/running-background-tasks-python.md [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+* Every task can send some meta information about its status or progress so far to Redis. This information can be useful for getting an insight into a task that runs for a long duration.
+
+Thanks for following along and I hope this post would have been useful for you.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I don't think this line adds anything.

**Review by sidhanthp — CHANGES_REQUESTED** [review]
I really enjoyed the style of the post and think you did a great job with a starter application, but I don't think the reader is left with a strong understanding of Redis Queue's, how to use them, and what's happening internally. 

I left you a couple comments, take a look at them and let me know what you think.

File: posts/running-background-tasks-python.md (hunk 1 above)
Comment (divyanshutomar) [reply posts/running-background-tasks-python.md]: It makes sense, I will rework on the introduction

File: posts/running-background-tasks-python.md (hunk 2 above)
Comment (divyanshutomar) [reply posts/running-background-tasks-python.md]: Thanks for pointing out the grammatical mistakes. 👍

File: posts/running-background-tasks-python.md (hunk 3 above)
Comment (divyanshutomar) [reply posts/running-background-tasks-python.md]: I think it's not about a small or big task. I feel 'dedicated task' will make more sense here?

I will add more details here explaining the role of messaging queues in microservices architecture.

File: posts/running-background-tasks-python.md (hunk 4 above)
Comment (divyanshutomar) [reply posts/running-background-tasks-python.md]: Noted

**Comment by divyanshutomar** [comment]
@sidhanthp I have made the required changes. Kindly review and let me know if anything else needs to be worked upon.

File: posts/running-background-tasks-python.md [hunk 15]
```diff
@@ -0,0 +1,185 @@
+---
+title: Running background tasks in Python using task queues
+author: Divyanshu Tomar
+---
+# Running background tasks in Python using task queues
+
+Let's say we have an e-commerce website where users can place orders for various products. Now there's a business requirement of finding out the kind of orders being placed and the most in-demand product in real-time. Also, for every order placed, the buyer should be conveyed of the confirmation of the order via an email or notification through a messaging service. In this case,if we process the order information on the same REST API service handling product requests, it may lead to significant problems. The API service may not able to respond in a short time as it can be blocked on the external services like the messaging service. This synchronous model worsens with a high number of orders being placed in a short span of time as the service may be preoccupied processing previous requests. Thus, these compute-intensive time-taking tasks like processing of orders on an e-commerce platform require an asynchronous approach.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I think this is too wordy. 

Before writing, I would think through what you want the reader to take away from the paragraph. As the first paragraph, you introduce the reader to the e-commerce website, but they've clicked on the article to learn about 'background tasks in Python' and I think it would be more effective starting with that. You only touch about the importance of background tasks at the end of the paragraph. 

Imagine I'm a reader who doesn't know what background tasks are / why I would want them, how would you introduce me with as little background?

File: posts/running-background-tasks-python.md [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+
+![architectire of web server and queue](./images/running-background-tasks-python/small-archi.png)
+
+Task or message queues are quite popular among microservices architecture. They enable each microservice to perform its dedicated task and work as a medium for inter-microservice communication. These queues store messages or data incoming from _producer_ microservices which can be processed or consumed by _consumer_ microservices. In the e-commerce example above, the REST API handling orders is a producer microservice which pushes these orders to the queue. Whereas, a data analysis microservice determining the kind of orders being placed or the messaging service can be considered a consumer microservice.
+
+## Queueing frameworks to the rescue
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I think this section would be more effective at the bottom. First, teach the reader why they need background tasks, about RQ, then point them to the other solutions.

File: posts/running-background-tasks-python.md [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+The execution of such tasks or jobs can be performed in the background by another process spawned for the sole purpose. These processes are usually called _workers_. They run concurrently with the main process (web server in our case) handling client requests. To list all the tasks which need to be executed, a job/task queue is maintained to store tasks along with its metadata created by incoming requests on the web server. The worker process then executes these tasks chronologically. This modular approach makes it easier for the web server to accommodate execution of such long-running tasks as it will not get blocked itself in doing so. This also means that the web server can respond to forthcoming client requests.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: Try to make this more concise. Here's an example --> 

`Workers` can be used to execute these tasks in the background. They run concurrently in the background using a queue, and the worker executes the tasks chronologically. 

This modular approach prevents the web server from being blocked from responding to incoming client requests.

File: posts/running-background-tasks-python.md [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+
+Redis Queue (RQ) is a python library which implements a simple job queue using Redis. It's a lightweight alternative that derives the good parts from existing queueing frameworks. A single worker process can be pointed to many queues to execute the tasks. We will be using this in the following application for its simplicity.
+
+
+There are many other queueing frameworks or services available. Refer [queues.io](http://queues.io/) to discover more.
+
+## Real World Application
+
+We will be writing a flask-based web application which retrieves _Goodreads_ book information like title, author, rating and description. The web server exposes an endpoint that accepts book URLs. A function will crawl and parse this URL  for meta information of the book. As this function will take time to execute and may lead to blocking of the main thread, we will execute it asynchronously by pushing it to Redis queue (RQ). RQ allows us to enqueue multiple function calls to a queue which can be executed parallelly by a separate worker process. It requires Redis server as a message broker for performing this operation. Let's get into the code and learn how we can use Redis queue in our web applications.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: Break the paragraph here. I think it makes it easier to follow. 

"""
... pushing it to Redis queue (RQ).

RQ allows us to enqueue ... 
"""

File: posts/running-background-tasks-python.md [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+  bsTree = BeautifulSoup(htmlString,"html.parser") # Build a searchable tree using fetched HTML
+  # Find the required book attributes in the tree
+  title = bsTree.find("h1", attrs={"id": "bookTitle"}).string
+  author = bsTree.find("a", attrs={"class": "authorName"}).span.string
+  rating = bsTree.find("span", attrs={"itemprop": "ratingValue"}).string
+  description = ''.join(bsTree.find("div", attrs={"id": "description"}).find("span", attrs={"style": "display:none"}).stripped_strings)
+  return dict(title=title.strip() if title else '',author=author.strip() if author else '',rating=float(rating.strip() if rating else 0),description=description)
+```
+
+We can now write a function called `parse_and_persist_book_info` that calls the above parsing function and persists the value to Redis so that it can be retrieved later. This function along with its arguments will be pushed to queue so that the worker process can execute it. Redis is a key-value store where the key should be unique else it may lead to overwriting of a previous value. Here `generate_redis_key_for_book` is a function that generates a unique key for a given book URL.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: Introduce a paragraph break here -->

"""
... process can execute it.

Redis is a key-value ...
"""

File: posts/running-background-tasks-python.md [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+
+```python
+# server.py
+import pickle
+
+#........
+#........
+
+# This generates a unique Redis key against a book URL
+generate_redis_key_for_book = lambda bookURL: 'GOODREADS_BOOKS_INFO:' + bookURL
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I would explain in a little more detail what you're doing here if this is a major piece of the post.

File: posts/running-background-tasks-python.md [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+
+def parse_and_persist_book_info(bookUrl):
+  redisKey = generate_redis_key_for_book(bookUrl) # get Redis key for given book URL 
+  bookInfo  = parse_book_link_for_meta_data(bookUrl) # get book meta information from parsing function above
+  # Set the value to Redis. Here pickle serializes the dictionary
+  redisClient.set(redisKey,pickle.dumps(bookInfo))
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I would explain `redisClient.set(redisKey,pickle.dumps(bookInfo))` further. I could be missing it, but I don't see where you initialized `redisClient`.

File: posts/running-background-tasks-python.md [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+```
+
+### The endpoint for accepting URLs
+
+Let's set up an endpoint that will accept a list of valid Goodreads book URLs. This is going to support POST method with URLs accepted as an array in _application/json_ body format. For validating the Goodreads book URLs we check for unique occurrences of URLs which starts with the string `https://www.goodreads.com/book/show/`. After the validation check, all valid URLs are pushed to Redis queue for parsing information. Here the method `enqueue_call` of Redis queue instance takes in a function that will be executed by worker process along with required arguments of the function.
```
Comment (sidhanthp) [inline posts/running-background-tasks-python.md]: I would take out the first couple sentences, and make the paragraph this -->
After the validation check, all valid URLs are pushed to Redis queue for parsing information. Here the method `enqueue_call` of Redis queue instance takes in a function that will be executed by worker process along with required arguments of the function.

I don't think the first couple sentences add to the explanation of Redis Queues.

**Review by sidhanthp — CHANGES_REQUESTED** [review]
Great job with the post, I've left you with a few comments that I think make the article a little easier to read and digest. 

I found it difficult to follow the longer paragraphs. Some of the paragraphs are big blocks of texts, which I think causes your eyes to glaze over the text. If you break up the paragraphs, or make the takeaways obvious using _italics_, I think it would make the article a little easier to follow.

**Comment by divyanshutomar** [comment]
@sidhanthp Updated

**Comment by divyanshutomar** [comment]
@sidhanthp Great job with the final touches. I guess we decided to have an example of a background job in the introductory para, that's why I kept the e-commerce example. Nevertheless, if you want to keep it concise the current state looks good to me.
